### DISCUSSION TITLE
Bind param of identifier type

### DISCUSSION THREAD

**Discussion by alecov** [sqlalchemy/sqlalchemy#9559] [Usage Questions]
Is it possible to bind a param of an identifier type?

For example, executing a `LISTEN` in PostgreSQL:

```python
# `Identifier` doesn't exist:
connection.execute(sql.text('LISTEN :channel').bindparams(sql.bindparam("channel", value=channel, type_=Identifier)))
```

**Comment by CaselIT** [comment]
Hi,

I'm not aware of a type called "Identifier" on postgresql. Do you have a documentation page for it? 

Also I'm not 100% sure listen allows parameters, but probably when using psycopg2 you can use it anyway since it does client side binding.
Have you tried not specifying the type, and let sqlalchemy infer it?

**Reply by alecov** [reply]
> Hi,
> 
> I'm not aware of a type called "Identifier" on postgresql. Do you have a documentation page for it?

Precisely, there isn't -- this is why I'm asking this question. That was an example of intended usage.

> Also I'm not 100% sure listen allows parameters, but probably when using psycopg2 you can use it anyway since it does client side binding. Have you tried not specifying the type, and let sqlalchemy infer it?

`LISTEN` doesn't accept a parameter in the ordinary sense (string, number, etc.), but it takes an identifier (a channel ID) to listen to. The idea is to pass this identifier parameter from a variable in a safe way. Also, there is no `pg_listen()` so there is no way to invoke a listen through a `SELECT`.

I'm sure there are other such situations as well, so I wonder whether SQLAlchemy has a trick to deal with them.

**Reply by zzzeek** [reply]
you'd want to start with the database driver you're using and the API the present for this .   For psycopg2, it looks like you can just emit SQL for LISTEN on the cursor as in the example at https://www.psycopg.org/docs/advanced.html?highlight=listen#asynchronous-notifications.    However they are not using a bound parameter, they are writing the SQL all as one string.  So if you can't use a parameter for psycopg2 cursor.execute() here, then you can't use it with SQLAlchemy either.  This is a driver-level detail.

**Reply by alecov** [reply]
>So if you can't use a parameter for psycopg2 cursor.execute() here, then you can't use it with SQLAlchemy either. This is a driver-level detail.

Makes sense.
